### ISSUE TITLE
SPIKE: role-based access control

### ISSUE THREAD

**Issue by AlexGilleran** [magda-io/magda#2038]
## Problem description
Right now not a lot of things are changed by users so we've gotten away with having admin users and not-admin users, but soon this will change and we'll need an access control system that supports more fine-grained permissions.

### A Rough Design
- A user can have many roles
- Roles can also have roles
- Roles have access to see/do certain things in the system
- Role membership is kept by the auth api
- What role a user needs to do a certain thing is baked into the service that does it
- Currently calls that go through the API get a JWT set on them that has the user id in them. This should change so that it has the users’s roles too. These should be stored in the session so that it doesn’t need to be manually looked up twice (get user id from session, get roles by user id from auth api) every time they make a call
    - Currently internal calls are made by passing a hard-coded user id that’s put into a signed JWT and attached to the call. These JWTs need to change too so that it has the right role, not just the right id.
   - This is a nice to have, but what happens if a users's roles change or a role's roles change, do we invalidate all those sessions?
- We should probably segregate the admin god account into several accounts… e.g. each api gets an account, and has role(s) to control what it does. E.g. adding things to the search index requires a certain role, which the indexer user has
- Roles can be added/removed by other users who themselves have the role that allows them to do so
- Roles can also be pre-populated from SSO providers
- Certain datasets will have certain roles attached to them in the registry (and hence search index) that determines who is allowed to edit/create/delete/view them

#### Potential problems
- How do we overcome roles having roles with a JWT? Do we put all user's roles plus all the roles that those roles have plus all the roles that those roles have etc in there?
- Same story for search - is it practical to just pass all the roles a user has to the search api and have it match against datasets?

## To Investigate:
- Are there any existing open source things or should we build something home-grown
- Protari has an access control thing, we should ask how they do it
- Someone needs to think through and enumerate all the roles that we need in the system
- Some kind of PoC to start from before we start splitting this into development tasks would be great.

**Comment by AlexGilleran** [comment]
@nahidakbar I know you were pretty interested in this

**Comment by t83714** [comment]
Just my 2 cents:
* How do we overcome roles having roles with a JWT? Do we put all user's roles plus all the roles that those roles have plus all the roles that those roles have etc in there?
  - We probably want to introduce the permission concept and forbid `* Roles can also have roles` . So that a user can have many different roles and each role can have many different permission. For SSO integration, we can need to create a AD group to magda role mapping
* Same story for search - is it practical to just pass all the roles a user has to the search api and have it match against datasets?
  - We probably need to store per resource permission somewhere. i.e. for a dataset, which groups / roles allow to have which permissions. This is the additional to the proposed functional level access control

**Comment by AlexGilleran** [comment]
We're way beyond this now!
